Update optimiser readme. (#6203)
ethereum/solidity: libyul/optimiser/README.md

The optimiser currently follows a purely greedy strategy and does not do any
backtracking.

In the following, all components of the optimiser are explained.
The main components of the optimiser are the following transformation steps:

 - [SSA Transform](#ssa-transform)
 - [Common Subexpression Eliminator](#common-subexpression-eliminator)
 - [Expression Simplifier](#expression-simplifier)
 - [Redundant Assign Eliminator](#redundant-assign-eliminator)

>  - [Full Function Inliner](#full-function-inliner)

Reviewer: In which category of the first list do these fit?

Author: Is that important?

Reviewer: Not really, I just didn't quite understand.

> Where I is a block that does not contain any function definitions (not even recursively)

Reviewer: Might wanna say that `I` can be empty.

and F is a list of function definitions such that no function contains a function definition.

### For Loop Init Rewriter

This transformation moves the initialization part of a for-loop to before
the loop:

    for { Init... } C { Post... } {
        Body...
    }

is transformed to

    {
        Init...
        for {} C { Post... } {
            Body...
        }
    }

This eases the rest of the optimisation process because we can ignore
the complicated scoping rules of the for loop initialisation block.

## Pseudo-SSA Transformation

The purpose of the following components is to get the program into a longer form,
so that other components can more easily work with it. The final representation
will be similar to a static-single-assignment (SSA) form, with the difference

> that it does not make use of explicit "phi" functions which combined the values

Reviewer: ```suggestion
that it does not make use of explicit "phi" functions which combines the values
```

"unused", "undecided" and "used" which signifies whether the assigned
value will be used later by a reference to the variable.

When an assignment is visited, it is added to the mapping in the "undecided" state
(see remark about for loops below) and every other assignment to the same variable
that is still in the "undecided" state is changed to "unused".
When a variable is referenced, the state of any assignment to that variable still
in the "undecided" state is changed to "used".

At points where control flow splits, a copy
of the mapping is handed over to each branch. At points where control flow
joins, the two mappings coming from the two branches are combined in the following way:
Statements that are only in one mapping or have the same state are used unchanged.
Conflicting values are resolved in the following way:

 - "unused", "undecided" -> "undecided"
 - "unused", "used" -> "used"
 - "undecided, "used" -> "used"

For for-loops, the condition, body and post-part are visited twice, taking
the joining control-flow at the condition into account.
In other words, we create three control flow paths: Zero runs of the loop,
one run and two runs and then combine them at the end.

> Running at most twice is enough because there are only three different states.

Author: It would be nice to have a proof for this bold statement...

Reviewer: "Since there are only three different states, two runs of the loop are clearly enough for convergence." q.e.d.

assignments or expression statements and all arguments have to be constants or variables.

The benefits of this form are that it is much easier to re-order the sequence of opcodes
and it is also easier to perform function call inlining. Furthermore, it is easier
to replace individual parts of expressions or re-organize the "expression tree".
The drawback is that such code is much harder to read for humans.

### SSA Transform

This stage tries to replace repeated assignments to
existing variables by declarations of new variables as much as
possible.
The reassignments are still there, but all references to the
reassigned variables are replaced by the newly declared variables.

Example:

    {
        let a := 1
        mstore(a, 2)
        a := 3
    }

is transformed to

    {
        let a_1 := 1
        let a := a_1
        mstore(a_1, 2)
        let a_3 := 3
        a := a_3
    }

Exact semantics:

For any variable ``a`` that is assigned to somewhere in the code (excluding
declarations with assignments) perform the following transforms:

>  - replace ``let a := v`` by ``let a_i := v   let a := a_i``

Reviewer: The sentence above just said `excluding declarations with assignments`

 - replace ``a := v`` by ``let a_i := v   a := a_i``
where ``i`` is a number such that ``a_i`` is yet unused.

Furthermore, always record the current value of ``i`` used for ``a`` and replace each
reference to ``a`` by ``a_i``.
The current value mapping is cleared for a variable ``a`` at the end of each block
in which it was assigned to and just after the for loop init block if it is assigned

> inside the for loop.

Reviewer: Inside the for loop init or body?

After this stage, the Redundant Assign Eliminator is recommended to remove the unnecessary
intermediate assignments.

This stage provides best results if the Expression Splitter and the Common Subexpression Eliminator

> is run right before it, because then it does not generate excessive amounts of variables.

Reviewer: ```suggestion
are run right before it, because then it does not generate excessive amounts of variables.
```

is side-effect free and its evaluation only depends on the values of variables
and the call-constant state of the environment. Most expressions are movable.
The following parts make an expression non-movable:

 - function calls (might be relaxed in the future)
 - opcodes that (can) have side-effects (like ``call`` or ``selfdestruct``)
 - opcodes that read or write memory, storage or external state information
 - opcodes that depend on the current PC, memory size or returndata size

### Dataflow Analyzer

The dataflow analyzer is not an optimizer step itself but is used as a tool
by other components. It traverses the AST and tracks the current value of
each variable, as long as that value is a movable expression.
It also tracks the variables that are part of the expression
that is currently assigned to each other variable. Upon each assignment to
a variable ``a``, the current stored value of ``a`` is updated and
all stored values of all variables ``b`` are cleared whenever ``a`` is part
of the currently stored expression for ``b``.

At control-flow joins, variables are cleared if they have or would be assigned
in any of the control-flow paths.

> 

Reviewer: Reviewed til here.

that variable. This is not a problem because such subexpressions have
to be movable.

All subexpressions that are identifiers themselves are replaced by their
current value if the value is an identifier itself.

The combination of the two rules above allow to compute a local value
numbering, which means that if two variables have the same
value, one of them will always be unused. The Unused Pruner or the
Redundant Assign Eliminator will then be able to fully eliminate such
variables.

This step is especially efficient if the expression splitter is run
before. If the code is in pseudo-SSA form,
the values of variables are available for a longer time and thus we
have a higher chance of expressions to be replaceable.

The expression simplifier will be able to perform better replacements
if the common subexpression eliminator was run right before it.

### Expression Simplifier

The Expression Simplifier uses the Dataflow Analyzer and makes use
of a list of equivalence transforms on expressions like ``X + 0 -> X``
to simplify the code.

It tries to match patterns like ``X + 0`` on each subexpression.
During the matching procedure, it resolves variables to their currently

> assigned expressions to be able to match more deeply nested patters

Reviewer: ```suggestion
assigned expressions to be able to match more deeply nested patterns
```

even when the code is in pseudo-SSA form.

Some of the patterns like ``X - X -> 0`` can only be applied as long

> as ``X`` is movable, because it would remove its potential side-effects.

Reviewer: ```suggestion
as ``X`` is movable, because otherwise it would remove its potential side-effects.
```

Since variable references are always movable, even if their current
value might not be, the Expression Simplifier is again more powerful
in split or pseudo-SSA form.

## Statement-Scale Simplifications

### Unused Pruner

This step removes the definitions of all functions that are never referenced.

It also removes the declaration of variables that are never referenced.
If the declaration assigns a value that is not movable, the expression is retained,
but its value is discarded.

> All movable expression statements are removed.

Reviewer: Unused expression statements?

 - replace if statement with true condition by its body
 - remove if statement with false condition
 - turn switch with single case into if
 - replace switch with only default case by ``pop(expression)`` and body
 - replace switch with literal expression by matching case body
 - replace for loop with false condition by its initialization part

This component uses the Dataflow Analyzer.

### Equivalent Function Combiner

If two functions are syntactically equivalent, while allowing variable
renaming but not any re-ordering, then any reference to one of the
functions is replaced by the other.

The actual removal of the function is performed by the Unused Pruner.

### Block Flattener

This stage eliminates nested blocks by inserting the statement in the
inner block at the appropriate place in the outer block:

    {
        let x := 2
        {
            let y := 3
            mstore(x, y)
        }
    }

is transformed to

    {
        let x := 2
        let y := 3
        mstore(x, y)
    }

As long as the code is disambiguated, this does not cause a problem because
the scopes of variables can only grow.

## Function Inlining

> ### Functional Inliner

Reviewer: ```suggestion
### Function Inliner
```

Author: It is different from the full inliner, but we might want to find a better name.

Reviewer: Partial function inliner? As opposed to Full..

Author: The thing is we would also have to rename the module...

Reviewer: Let's keep Functional Inliner then, not a big deal.

> The functional inliner performs function inlining such that the result of the inlining is a

Reviewer: ```suggestion
The result of function inlining is a
```

Author: I think this would make it too simple. This sentence differentiates the functional inliner from the full inliner. I'll try to find a better wording...

It does not make use of any information concerning the commutability of opcodes;
if moving the value of a variable to its place of use would change the order
of any function call or opcode execution, the transformation is not performed.

Note that the component will not move the assigned value of a variable assignment
or a variable that is referenced more than once.

The snippet ``let x := add(0, 2) let y := mul(x, mload(2))`` is not transformed,
because it would cause the order of the call to the opcodes ``add`` and
``mload`` to be swapped - even though this would not make a difference
because ``add`` is movable.

When reordering opcodes like that, variable references and literals are ignored.
Because of that, the snippet ``let x := add(0, 2) let y := mul(x, 3)`` is
transformed to ``let y := mul(add(0, 2), 3)``, even though the ``add`` opcode
would be executed after the evaluation of the literal ``3``.

### SSA Reverser

This is a tiny step that helps in reversing the effects of the SSA transform
if it is combined with the Common Subexpression Eliminator and the
Unused Pruner.

The SSA form we generate is detrimental to code generation on the EVM and

> WebAssembly alike because it generatse many local variables. It would

Reviewer: ```suggestion
WebAssembly alike because it generates many local variables. It would
```

be better to just re-use existing variables with assignments instead of
fresh variable declarations.

The SSA transform rewrites

    a := E
    mstore(a, 1)

to

    let a_1 := E
    a := a_1
    mstore(a_1, 1)

The problem is that instead of ``a``, the variable ``a_1`` is used
whenever ``a`` was referenced. The SSA transform changes statements
of this form by just swapping out the declaration and the assignment. The above
snippet is turned into

    a := E
    let a_1 := a
    mstore(a_1, 1)

This is a very simple equivalence transform, but when we now run the
Common Subexpression Eliminator, it will replace all occurrences of ``a_1``
by ``a`` (until ``a`` is re-assigned). The Unused Pruner will then
eliminate the variable ``a_1`` altogether and thus fully reverse the
SSA transform.

### Stack Compressor

> One problem that makes code generation for the Ethereum Virtual Machine

Reviewer: EVM was already used, could change to
```suggestion
One problem that makes code generation for the EVM
```

hard is the fact that there is a hard limit of 16 slots for reaching
down the expression stack. This more or less translates to a limit
of 16 local variables. The stack compressor takes Yul code and
compiles it to EVM bytecode. Whenever the stack difference is too
large, it records the function this happened in.

> For each function that caused such a problem, the Rematerialiser

Reviewer: `Rematerialiser` sounds like it's from Star Trek :p
